### PR TITLE
🗄️ Easier iterating of and searching in archives 

### PR DESCRIPTION
**Pull request by rowanseymour**

### CODE REVIEW

File: temba/archives/models.py [hunk 1]
```diff
@@ -1,64 +1,66 @@
...
 
 
 class Archive(models.Model):
     DOWNLOAD_EXPIRES = 60 * 60 * 24  # Up to 24 hours
+
     TYPE_MSG = "message"
     TYPE_FLOWRUN = "run"
-
-    PERIOD_MONTHLY = "monthly"
-    PERIOD_DAILY = "daily"
```
Comment (rowanseymour) [inline temba/archives/models.py:21]: these are repeated below

File: temba/archives/migrations/0012_auto_20200807_1913.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
+# Generated by Django 2.2.10 on 2020-08-07 19:13
+
+import django.db.models.deletion
+import django.utils.timezone
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
```
Comment (rowanseymour) [inline temba/archives/migrations/0012_auto_20200807_1913.py:8]: just removing a bunch of unused help_text

File: temba/archives/management/commands/search_archives.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+from temba.utils import json
+
+
+class Command(BaseCommand):  # pragma: no cover
+    help = "Searches through archives"
+
+    def add_arguments(self, parser):
+        parser.add_argument("org_id", help="ID of the org whose archives will be searched")
+        parser.add_argument("archive_type", choices=[Archive.TYPE_MSG, Archive.TYPE_FLOWRUN])
+        parser.add_argument("--expression", type=str, action="store", dest="expression", default="")
```
Comment (rowanseymour) [inline temba/archives/management/commands/search_archives.py:14]: was using this to test S3 Select stuff but have convinced myself it's useful enough to keep around

File: temba/msgs/models.py [hunk 4]
```diff
@@ -1607,63 +1607,45 @@ def write_export(self):
...
+                visibility = "visible"
+                if system_label:
+                    visibility, direction, msg_type, statuses = SystemLabel.get_archive_attributes(system_label)
 
-                    if created_on < start_date or created_on > end_date:  # pragma: can't cover
+                    if record["direction"] != direction:
                         continue
 
-                    if group_contacts and record["contact"]["uuid"] not in group_contacts:
+                    if msg_type and record["type"] != msg_type:
```
Comment (rowanseymour) [inline temba/msgs/models.py:1633]: lots of stuff here could be part of an S3 select query if we want to make that a hard dependency

File: temba/archives/models.py (hunk 1 above)
Comment (nicpottier) [reply temba/archives/models.py:21]: So nice, I'll say it twice!

File: temba/archives/models.py [hunk 5]
```diff
@@ -118,20 +120,90 @@ def get_download_link(self):
...
+        if expression:
+            response = s3.select_object_content(
+                **self.s3_location(),
+                ExpressionType="SQL",
+                Expression=f"SELECT * FROM s3object s WHERE {expression}",
+                InputSerialization={"CompressionType": "GZIP", "JSON": {"Type": "LINES"}},
+                OutputSerialization={"JSON": {"RecordDelimiter": "\n"}},
+            )
+
+            for event in response["Payload"]:
```
Comment (nicpottier) [inline temba/archives/models.py:187]: Does s3 select have any maximum number of rows it returns?

Comment (rowanseymour) [reply temba/archives/models.py:187]: no, and to be clear `response["Payload"]` is a stream which is fetching on demand https://botocore.amazonaws.com/v1/documentation/api/latest/reference/eventstream.html

**Pull request merged**